invenio_rest/csrf.py

```diff
@@ -0,0 +1,262 @@
+            current_app.config['CSRF_ALLOWED_CHARS'],
+        )
+    )
+    return encoded_token
+
+
+def _get_csrf_token():
+    try:
+        csrf_cookie = request.cookies[
+            current_app.config['CSRF_COOKIE_NAME']]
+    except KeyError:
+        return None
+    return _decode_csrf(csrf_cookie)
+
+
+def _decode_csrf(data):
+    csrf_serializer = _get_csrf_serializer()
+    try:
+        return csrf_serializer.loads(data)
+    except BadSignature:
+        raise RESTCSRFError()
+
+
+def _set_token(response):
+    response.set_cookie(
+        current_app.config['CSRF_COOKIE_NAME'],
+        _get_new_csrf_token(),
+        max_age=current_app.config.get(
+            'CSRF_COOKIE_MAX_AGE', 60*60*24*7*52),
+        domain=current_app.config.get(
+            'CSRF_COOKIE_DOMAIN',
+            current_app.session_interface.get_cookie_domain(
+                current_app)),
+        path=current_app.session_interface.get_cookie_path(
+                current_app),
+        secure=current_app.config.get('SESSION_COOKIE_SECURE', False),
+        httponly=False,
+        samesite=current_app.config['CSRF_COOKIE_SAMESITE'],
+    )
+
+
+def _get_submitted_csrf_token():
+    header_name = current_app.config['CSRF_HEADER']
+    csrf_token = request.headers.get(header_name)
+    if csrf_token:
+        return csrf_token
+    return None
+
+
+def _is_referrer_secure(referrer):
+    return 'https' in referrer.scheme or \
+        not current_app.config['CSRF_FORCE_SECURE_REFERRER']
+
+
+def _reject(reason):
+        abort(400, reason)
+
+
+def csrf_validate():
+    """Check csrf cookie against request headers."""
+    if request.method not in ('GET', 'HEAD', 'OPTIONS', 'TRACE'):
+        if request.is_secure:
+            referrer = request.referrer
+
+            if referrer is None:
+                return _reject(REASON_NO_REFERER)
+
+            referrer = urlparse(referrer)
+            # Make sure we have a valid URL for Referer.
+            if '' in (referrer.scheme, referrer.netloc):
+                return _reject(REASON_MALFORMED_REFERER)
+
+            # Ensure that our Referer is also secure.
+            if not _is_referrer_secure(referrer):
+                return _reject(REASON_INSECURE_REFERER)
+
+            if referrer.hostname not in current_app.config.get(
+                                            'APP_ALLOWED_HOSTS'):
+                reason = REASON_BAD_REFERER % referrer.geturl()
+                return _reject(reason)
+
+        csrf_token = _get_csrf_token()
+        if csrf_token is None:
+            return _reject(REASON_NO_CSRF_COOKIE)
+
+        request_csrf_token = _get_submitted_csrf_token()
+        if not request_csrf_token:
+            _reject(REASON_BAD_TOKEN)
+
+        decoded_request_csrf_token = _decode_csrf(request_csrf_token)
+
+        if csrf_token != decoded_request_csrf_token:
+            return _reject(REASON_BAD_TOKEN)
+
+
+def reset_token(self):
```

Comment:
when is this method called? We could simply reset it on each `POST` and `PUT` (maybe also `PATH` and `DELETE`) so we don't have this extra state and the code is more simple.
See: https://jcbaey.com/authentication-in-spa-reactjs-and-vuejs-the-right-way (find in the page`use PUT or POST`)

Reply:
This method is meant to be used in `invenio-accounts` whenever a user is logging in...This is just to refresh the token explicitly whenever the user logs  in/out.

Reply:
There's something similar in Django's implementation, which mentions if the CSRF token was used or not during the request... But it's not reset on every POST/PUT/DELETE/... request

Reply:
I would say that the CSRF should be not related to login/logout/user session... so unless it is expensive to generate the CSRF token, I would just do it on POST/PUT... client implementation should be completely transparent to this, just take from the cookie and add to the header, without even checking it?

Reply:
So, that was our first implementation in which we were generating a random value after each request(POST, PUT etc.) but we said to do it on demand because of the jwt payload. We were thinking that users typically will put there information after accessing db most probably. But now we removed completely the jwt payload and the csrf cookie is a pure random encoded value. So I guess @slint we are fine if we go back to the initial implementation? @ntarocco generating a new value after each request would be ok right?

Reply:
So decided to keep this, and change the if to re-generate the token to something like:
`if vulnerable_method OR csrf_cookie_needs_reset: reset()`